BridgeSDK/BridgeAPI/BridgeAPITypes/SBBDateTimeRangeResourceList.m

```diff
@@ -55,17 +55,29 @@ - (void)reconcileWithDictionaryRepresentation:(NSDictionary *)dictionary objectM
     // as well, since that most likely means there was some kind of problem that required cleanup on the
     // server and the missing ones are no longer canonically valid. We'll look at the startTime and endTime
     // and remove our cached items in that date range (inclusive on the lower bound, exclusive on the upper),
-    // then add in the newly-retrieved ones, and sort by scheduledOn before adding back to
-    // the cached object's items[].
+    // then add in the newly-retrieved ones, and sort by date(s) before setting back as the cached object's items[].
     
-    NSString *comparisonKey = NSStringFromSelector(@selector(scheduledOn));
-    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K < %@ OR %K >= %@",
-                              comparisonKey, self.startTime,
-                              comparisonKey, self.endTime];
+    NSString *scheduledOnKey = NSStringFromSelector(@selector(scheduledOn));
+    NSString *expiresOnKey = NSStringFromSelector(@selector(expiresOn));
+    
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(%K < %@ && %K != nil && %K < %@) OR %K >= %@",
+                              scheduledOnKey, self.startTime,
+                              expiresOnKey,
+                              expiresOnKey, self.startTime,
+                              scheduledOnKey, self.endTime];
```

Comment:
You seem to be missing the finished on part of this. You also need to filter on `finishedOn`, don't you? 

````
NSPredicate *finishedOutsideRange = [NSPredicate predicateWithFormat:@"%K != nil AND (%K < %@ OR %K >= %@)", finishedOnKey, finishedOnKey, self.startTime, finishedOnKey, self.endTime ];
NSPredicate *scheduledOutsideRange = [NSPredicate predicateWithFormat:@"(%K < %@ && %K != nil && %K < %@) OR %K >= %@",
                              scheduledOnKey, self.startTime,
                              expiresOnKey,
                              expiresOnKey, self.startTime,
                              scheduledOnKey, self.endTime];
NSPredicate *predicate = [NSCompoundPredicate orPredicateWithSubpredicates: @[scheduledOutsideRange, finishedOutsideRange]];
````

You might want to take a look at the predicates that I created in an extension on `SBBScheduledActivity`. Specifically, this one:

````
    public static func scheduledPredicate(on date:Date) -> NSPredicate {
        let startOfDay = date.startOfDay()
        let startOfNextDay = startOfDay.addingNumberOfDays(1)
        
        // Scheduled for this date or prior
        let scheduledKey = #keyPath(scheduledOn)
        let scheduledThisDayOrBefore = NSPredicate(format: "%K == nil OR %K < %@", scheduledKey, scheduledKey, startOfNextDay as CVarArg)
        let unfinished = unfinishedPredicate()
        let finishedOnThisDay = finishedPredicate(on: date)
        
        let expiredKey = #keyPath(expiresOn)
        let expiredOnThisDay = NSPredicate(format: "%K == nil OR (%K >= %@ AND %K < %@)", expiredKey, expiredKey, startOfDay as CVarArg, expiredKey, startOfNextDay as CVarArg)
        let expiredOnOrAfterThisDay = NSPredicate(format: "%K == nil OR %K > %@", expiredKey, expiredKey, startOfDay as CVarArg)
        
        switch(startOfDay.compare(Date().startOfDay())) {
            
        case .orderedAscending:
            // build a filter for a day in the past that includes expired on that day OR completed on that day
            let expired = NSCompoundPredicate(andPredicateWithSubpredicates: [unfinished, expiredOnThisDay])
            let finishedOrExpired = NSCompoundPredicate(orPredicateWithSubpredicates: [finishedOnThisDay, expired])
            return NSCompoundPredicate(andPredicateWithSubpredicates: [scheduledThisDayOrBefore, finishedOrExpired])
        
        case .orderedSame:
            // build a filter for today that includes activites completed today, expiring today or later and scheduled to 
            // include today
            let unfinishedOrFinishedToday = NSCompoundPredicate(orPredicateWithSubpredicates: [unfinished, finishedOnThisDay])
            return NSCompoundPredicate(andPredicateWithSubpredicates: [scheduledThisDayOrBefore, unfinishedOrFinishedToday, expiredOnOrAfterThisDay])
        
        case .orderedDescending:
            // For the future, we only want unfinished schedules
            return NSCompoundPredicate(andPredicateWithSubpredicates: [scheduledThisDayOrBefore, unfinished, expiredOnOrAfterThisDay])
        }
    }
````